Initial proof of concept (#1)
collective/collective.indexvocabularies: README.rst

A Plone addon for creating dynamic vocabularies from catalog indexes.

> This package creates dynamic vocabularies based on any index in plone.

Reviewer: ```suggestion
This package allows an admin to create dynamic named vocabularies from data within the plone site.
Vocabularies can then be used in metadata fields of your content types or in facets on a search block in volto.
```

> Typical use case for this is to provide a 'tags' type field.

Reviewer: ```suggestion
Currently supported is making vocabularies from existing unique values of a metadata field. This is useful to create
a custom Tags field which is addable by editors, similar to the built in Subjects field.
```

A Plone addon for creating dynamic vocabularies from catalog indexes.

> This package allows an admin to create dynamic named vocabularies from data within the plone site.

Reviewer 2: ```suggestion
This package allows an admin to create dynamically named vocabularies from data within the Plone site.
```

> Vocabularies can then be used in metadata fields of your content types or in facets on a search block in volto.

Reviewer 2: ```suggestion
Vocabularies can then be used in metadata fields of your content types or in facets on a search block in Volto.
```

> Currently supported is making vocabularies from existing unique values of a metadata field. This is useful to create

Reviewer 2: ```suggestion
Currently supported is making vocabularies from existing unique values of a metadata field.
```

> a custom Tags field which is addable by editors, similar to the built in Subjects field.

Reviewer 2: ```suggestion
This is useful to create a custom ``Tags`` field which is addable by editors, similar to the built in ``Subjects`` field.
```

and then running ``bin/buildout``

Activation
----------

Activate the add-on by installing it in the Addons Control panel.

Configure
---------

> To create a vocabulary from an index, go to the 'Index Vocabularies' control
> panel and select the indexes you wish to create vocabularies for.
> 
> To use the vocabulary in a tag field, you will need to edit the XML Schema of
> your content type. Currently this is only possible via the Classic UI
> (Control Panel -> Content Types -> Schema -> Edit XML Schema)

Reviewer: ```suggestion

To create a custom tag field on a content type you will need to do the following

1. Create a new field of type "Multiselect"
2. Save the content type. This will create an index for this field.
3. Go to Dynamic Vocabularies in the control panel
4. Select the Index matching the field you just created and click "Make Vocabulary".
5. Copy the name of the vocabulary generated.
8. Edit your content type again but this time via the editing it's XML
   - Currently this is only possible via the Classic UI (Control Panel -> Content Types -> Schema -> Edit XML Schema)
9. Change the widget to a TagSelectWidget with the vocabulrary name that was generated.  

 e.g. 
    <field name="test" type="zope.schema.Set">
      <title>test</title>
      <value_type type="zope.schema.TextLine" />
      <form:widget type="collective.indexvocabularies.widgets.TagSelectWidget">
          <vocabulary>collective.dynamicvocabularies.uniquevalues.test</vocabulary>
      </form:widget>
    </field>
  
```

Reviewer: why does this include our own widget? The standard one should work right?

Configure
---------

> To create a custom tag field on a content type you will need to do the following

Reviewer 2: ```suggestion
To create a custom tag field on a content type, you will need to do the following.
```

> 1. Create a new field of type "Multiselect"

Reviewer 2: ```suggestion
1. Create a new field of type ``Multiselect``.
```

2. Save the content type. This will create an index for this field.

> 3. Go to Dynamic Vocabularies in the control panel

Reviewer 2: ```suggestion
3. Go to Dynamic Vocabularies in the control panel.
```

> 4. Select the Index matching the field you just created and click "Make Vocabulary".

Reviewer 2: ```suggestion
4. Select the Index matching the field you just created, and click "Make Vocabulary".
```

5. Copy the name of the vocabulary generated.

> 8. Edit your content type again but this time via the editing it's XML

Reviewer 2: ```suggestion
6. Edit your content type again, but this time by editing its XML.
```

>    - Currently this is only possible via the Classic UI (Control Panel -> Content Types -> Schema -> Edit XML Schema)

Reviewer 2: ```suggestion
   Currently this is only possible via the Classic UI (Control Panel -> Content Types -> Schema -> Edit XML Schema).
```

> 9. Change the widget to a TagSelectWidget with the vocabulrary name that was generated.  

Reviewer 2: ```suggestion
7. Change the widget to a ``TagSelectWidget`` with the vocabulary name that was generated.  
```

>  e.g. 

Reviewer 2: ```suggestion
   .. code-block:: xml
```

Reviewer 2: You might have to fiddle with indentation to get the XML to render with proper syntax highlighting. In general, I always indent my enumerated lists and code-blocks with 4 spaces. Your enumerated lists are indented with only 3 spaces, which can cause funky rendering in reStructuredText.

 - A persistent ``IVocabularyFactory`` is registered with the name
   ``collective.indexvocabularies.{index_name}``
 - A series of registry entries are created that register the widget as a
   facet and filter for ``plone.app.querystring``

The addon also subclasses the `default IAjaxSelect tag widget <https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/interfaces.py#L82>`_
- from `plone.app.z3cform <https://github.com/plone/plone.app.z3cform>`_ in
order to provide supermodel import/export support. This could easily be added to
``plone.app.z3cform`` which would remove the need for an additional widget.

Alternatives
------------

Depending on your specific usecase you might be able to try the following:

 - Create custom behaviors in your own addon

>  - Use the `plone.app.vocabularies.Catalog` vocabualary. See `this discussion <https://community.plone.org/t/widget-parameter-for-catalogsource-based-choicefield/18129/3>`_

Reviewer 2: reStructuredText uses double backticks for inline literals, whereas Markdown uses single backticks.
```suggestion
 - Use the ``plone.app.vocabularies.Catalog`` vocabualary. See `this discussion <https://community.plone.org/t/widget-parameter-for-catalogsource-based-choicefield/18129/3>`_
```
